### ISSUE TITLE
expanding dplyr episode

### ISSUE THREAD

**Issue by marcianito** [swcarpentry/r-novice-gapminder#255]
Since this "new grammar" [of dyplyr] gained wide attention and more and more usage throughout the R-Community, I think this is a great and important episode!
It is already a good overview of dplyr's basis functions but I feel like **introducing the concept of joins** would be highly beneficial to the learners. This would give opportunities to merge, limit and expand different datasets. Furthermore I think it would also be beneficial in preparation of data to be plotted with ggplot.

A smaller comment: What about pointing to the use of **ifelse() in combination with mutate**!? I think this would be of much use for people starting with dplyr and "transforming existing work flows to this grammar".

What do you people think?

**Comment by naupaka** [comment]
Hi @marcianito, thanks for the comment. 

The lessons are already quite full, and it's always tempting to want to add more material. I think the challenge with talking about joins in this context is that we are working with a single dataset - gapminder - and so it doesn't lend itself easily to doing a join unless we introduce an additional dataset to join to, and then that adds more cognitive overhead and etc. So unless you can think of a straightforward way to introduce that given what else is already in the lesson, I think it might be too much (although I agree joins are hugely useful). 

As for `ifelse()` plus `mutate()`, I think that that might be easier to add in - do you have a sense for a good place where it might fit and how you would introduce it? PR would be welcome and we could talk through how best to integrate that.

**Comment by marcianito** [comment]
Hi @naupaka,

Yes, I agree that the lessons are quite full and I understand and share your opinion about the gapminder dataset. One option to integrate joins could be in the part of "combining dplyr and ggplot2", where the dataset to plot could be expanded by a new column with extra information wanted for plotting. This new column could be created in a data.frame apart, including the plotting information / categorizations and could then be joined with the prepared gapminder dataset.
For example:

```R
# create data.frame with dominant religion
ReligionPerCountry = data.frame(country = c("Afghanistan", ...),
dominantReligion = c("Muslim", ..) )
```
```R
# join information about religion with gapminder data for plotting
gapminder %>% 
# Get the start letter of each country 
mutate(startsWith = substr(country, start = 1, stop = 1)) %>% 
# Filter countries that start with "A" or "Z"
filter(startsWith %in% c("A", "Z")) %>%
# Join data with information about religion
left_join(ReligionPerCountry) %>%
# Make the plot
ggplot(aes(x = year, y = lifeExp, color = dominantReligion)) + 
geom_line() + 
facet_wrap( ~ country)
```

**Comment by marcianito** [comment]
With respect to `iflese()` and `mutate()`, I think a good (and short) way could be right after the "using mutate()" block, adding a block with something like:

## Combining mutate() and logical filtering
The creation of new variables with `mutate()` can be combined with logical filtering using `ifelse()`. This way the values of the new variables depend on the logical expression given in the `ifelse()`-statement:
```R
# calculating gdp only when life expectation exceeds 30 years,
# but conserving all data for plotting the complete dataset
gdp_pop_bycontinents_byyear <- gapminder %>%
    mutate(gdp_billion= ifelse(lifeExp >= 30, gdpPercap*pop/10^9, NA) %>%
    group_by(continent,year) %>%
    summarize(mean_gdpPercap=mean(gdpPercap, na.rm = T),
              sd_gdpPercap=sd(gdpPercap, na.rm = T),
              mean_pop=mean(pop, na.rm = T),
              sd_pop=sd(pop, na.rm = T),
              mean_gdp_billion=mean(gdp_billion, na.rm = T),
              sd_gdp_billion=sd(gdp_billion, na.rm = T))
```
What do you think about this? It would be short, only slightly modify the example given in the introduction of `mutate()` but give the learner the knowledge of an important way how to "categorize" newly created variables depending on existing data of the dataset used.

**Comment by naupaka** [comment]
Hi @marcianito I would be happy to look over a PR to add the chunk on using `mutate()` and `ifelse()` together. It's a handy trick that doesn't get the lesson too far off track.

I think unfortunately that adding something in about joins is out of scope at this point - there are lots of other great resources on the web for people to learn this and I think it's important the lessons don't get too over-stuffed.

**Comment by naupaka** [comment]
There's also `mutate_if()`, which can function similarly to `mutate(ifesle(...))`.

**Comment by marcianito** [comment]
@naupaka Sounds good!
I will have a look and initiate the PR.

**Comment by naupaka** [comment]
Closed by #271
